### PR TITLE
VDS-2583  VDS-2573 [Import descriptions] Implement Update/Create. Statistics Page

### PR DESCRIPTION
**Pull request by trueboroda**
Description: 

--

QA-test: VDS-2586

Demo-test: VDS-2587

Download artifact URL: https://vc3prerelease.blob.core.windows.net/packages/VirtoCommerce.DescriptionExportImport_1.0.0-pr-8-2b83a860.zip

### CODE REVIEW

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Validation/ImportReviewValidator.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                .WithImportState()
+                .DependentRules(() =>
+                    RuleFor(x => x.Record.ReviewType)
+                        .Must((_, reviewType, context) =>
+                        {
+                            var reviewTypes = (string[])context.ParentContext.RootContextData[
+                                catalogCore.ModuleConstants.Settings.General.EditorialReviewTypes.Name];
+                            return reviewTypes.Contains(reviewType, StringComparer.InvariantCultureIgnoreCase);
+                        })
+                        .WithInvalidValueCodeAndMessage("")
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Validation/ImportReviewValidator.cs]: Is here don't need the message?

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Validation/ImportReviewValidator.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                .WithImportState()
+                .DependentRules(() =>
+                        RuleFor(x => x.Record.LanguageCode)
+                            .Must((_, language, context) =>
+                            {
+                                var languages = (string[])context.ParentContext.RootContextData[
+                                        coreModuleCore.ModuleConstants.Settings.General.Languages.Name];
+                                return languages.Contains(language, StringComparer.InvariantCultureIgnoreCase);
+                            })
+                            .WithInvalidValueCodeAndMessage("")
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Validation/ImportReviewValidator.cs]: Is here don't need the message?

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                                            importProgress.UpdatedCount;
+            }
+        }
+
+
+        private static void SetIdAndSkuToNullByExistence(ImportRecord<CsvEditorialReview>[] importReviewRecords, EditorialReview[] existedReviews)
+        {
+            foreach (var importRecord in importReviewRecords)
+            {
+                if (existedReviews.Any(x => x.Id == importRecord.Record.Id))
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs]: EqualsInvariant?

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                }
+            }
+        }
+
+        private static void AddReviewsToProducts(IList<CatalogProduct> productsForReviewAdding, CsvEditorialReview[] importReviewsForAdding)
+        {
+            foreach (var product in productsForReviewAdding)
+            {
+                var productImportReviews =
+                    importReviewsForAdding.Where(x => x.ProductSku.EqualsInvariant(product.Code)).ToArray();
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs]: Maybe put this LINQ into the foreach statement

File: src/VirtoCommerce.DescriptionExportImportModule.Data/VirtoCommerce.DescriptionExportImportModule.Data.csproj [hunk 5]
```diff
@@ -1,4 +1,4 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/VirtoCommerce.DescriptionExportImportModule.Data.csproj:1]: UTF8 with BOM? :) Maybe revert?

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs (hunk 4 above)
Comment (trueboroda) [reply src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs]: How?

Comment (trueboroda) [reply src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs]: These are different collections.

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+                foreach (var productImportReview in productImportReviews)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs]:
```suggestion
                foreach (var productImportReview in productImportReviews.Where(x => x.ProductSku.EqualsInvariant(product.Code)))
```

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs (hunk 4 above)
Comment (pushnitsa) [reply src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs]:
```suggestion
                    
```

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                    importRecord.Record.Id = null;
+                }
+            }
+        }
+
+        private static void AddReviewsToProducts(IList<CatalogProduct> productsForReviewAdding, CsvEditorialReview[] importReviewsForAdding)
+        {
+            foreach (var product in productsForReviewAdding)
+            {
+                var productImportReviews =
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/CsvPagedEditorialReviewDataImporter.cs]:
```suggestion

```

File: src/VirtoCommerce.DescriptionExportImportModule.Data/VirtoCommerce.DescriptionExportImportModule.Data.csproj (hunk 5 above)
Comment (trueboroda) [reply src/VirtoCommerce.DescriptionExportImportModule.Data/VirtoCommerce.DescriptionExportImportModule.Data.csproj:1]: Now file has encoding UTF without BOM. It means the file had UTF-8 with BOM before changes.

**Review by pushnitsa — APPROVED** [review]

**Review by Pav-el — APPROVED** [review]

**Pull request merged**